R/generate_cmp_tbl.R

```diff
@@ -222,6 +222,181 @@ generate_lipidblast_tbl <- function(json_path){
 }
 
 
+#' @title Generate a table with HMDB compounds
+#'
+#' @description `generate_hmdb_tbl` processes one or more HMDB xml files
+#'     (downloaded from http://http://www.hmdb.ca/) or an HMDB file in SDF
+#'     format and extracts general compound information.
+#'
+#' @note At present only a subset of the available data provided by HMDB are
+#'     extracted.
+#' 
+#' @param file `character` with the name(s) of xml files downloaded from HMDB,
+#'     or the file name of a HMD file in SDF format.
+#'     For xml files: can be a single xml file (containing all compounds from
+#'     an HMDB release) or the names of several xml files, each supposed to
+#'     provide data for one compound.
+#'
+#' @return A `tbl` with general compound information (one row per compound):
+#' + `id`: the HMDB ID of the compound.
+#' + `name`: the compound's name.
+#' + `inchi`: the inchi of the compound.
+#' + `formula`: the chemical formula.
+#' + `mass`: the compound's mass. The value from HMDB's
+#'   `monisotopic_molecular_weight` is returned.
+#'
+#' @importFrom dplyr as.tbl
+#' @importFrom tools file_ext
+#'
+#' @export
+#' 
+#' @md
+#'
+#' @export
+#'
+#' @author Johannes Rainer
+#'
+#' @examples
+#'
+#' ## Extract compound information from a single xml file from HMDB.
+#' fl <- system.file("extdata/hmdb/HMDB0000001.xml", package = "PeakABro")
+#' hmdb <- generate_hmdb_tbl(fl)
+#' colnames(hmdb)
+#' hmdb
+#' 
+#' ## Extract compound information from a (subset of a) HMDB release xml file.
+#' fl <- system.file("extdata/hmdb/hmdb_sub.xml", package = "PeakABro")
+#' hmdb <- generate_hmdb_tbl(fl)
+#' colnames(hmdb)
+#' hmdb
+#'
+#' ## Extract compound information from a HMDB file in SDF format
+#' fl <- system.file("extdata/hmdb/hmdb_sub.sdf", package = "PeakABro")
+#' hmdb <- generate_hmdb_tbl(fl)
+#' hmdb
+#' 
+#' ## Extract the masses
+#' hmdb$mass
+generate_hmdb_tbl <- function(file) {
+    if (missing(file))
+        stop("'file' is required")
+    file <- normalizePath(file)
+    if (any(!file.exists(file)))
+        stop("Can not find file: ", paste(file[!file.exists(file)],
+                                          collapse = ", "))
+    ## Check if we've got an xml file or a sdf file...
+    if (any(tolower(file_ext(file)) == "xml")) {
+        res <- as.tbl(do.call(rbind, lapply(file, simple_parse_hmdb_xml)))
```

Comment:
Alternative:

```r
file %>% 
map(simple_parse_hmdb_xml) %>% 
bind_rows
```

Reply:
What's the benefit from that? Is it faster?

Reply:
Very few I think :D 
Only the tidyverse ethos of reading forward instead of inside out.
It is not important really.

It is slightly safer and according to the doc also more efficient.
bind_rows doesn't allow columns of different types and binds by column name instead of position.

I think you could even use `map_dfr` here which would make it a single operation.

Reply:
Cool. I'll check that.

Reply:
Nice. I like that. But I had to add a `as.tbl`.